### PR TITLE
Smart stepping implemented for Groovy. SmartSteppingFilterWrapper int…

### PR DESCRIPTION
**Pull request by ebarboni**
…roduced as a step-specific filter.

#4004 second merge attempt

### CODE REVIEW

**Comment by ebarboni** [comment]
@entlicher recreated #4004 from patch and reproposed PR as we had issue. I you prefer resubmitting by yourself feel free to close this one

File: groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                return StopOrStep.stop();
+            }
+        } else {
+            logger.fine(" => FINISHED (no Groovy), stop");
+            return StopOrStep.stop();
+        }
+    }
+
+    private static boolean isPackageToSkip(String className) {
+        for (String pattern : PATTERNS_SKIP_IN_GROOVY) {
```
Comment (lbownik) [inline groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java:97]: How about the following ?

return Stream.of(PATTERNS_SKIP_GROOVY).anyMatches(className::startsWith);

File: groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        return true;
+    }
+
+    @Override
+    public StopOrStep stopAt(ContextProvider lookupProvider, CallStackFrame frame, SmartSteppingFilter filter) {
+        String className = frame.getClassName();
+        if (logger.isLoggable(Level.FINE)) {
+            logger.fine("GroovySmartStepping.stopAt("+className+")");
+        }
+        JPDAThread thread = frame.getThread();
```
Comment (lbownik) [inline groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java:67]: "thread" seems to be unused

File: groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        boolean inGroovy = false;
+        for (String gp : GROOVY_PACKAGES) {
```
Comment (lbownik) [inline groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java:69]: how about
if(Stream.of(GROOVY_PACKAGES).anyMatch(className::startsWith) {
    steppingInGroovy.put(filter, true);
    logger.fine(" => In Groovy: Step In");
    return StopOrStep.step(0, JPDAStep.STEP_INTO);
}

File: groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java (hunk 2 above)
Comment (neilcsmith-net) [reply groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java:67]: @lbownik this code was already reviewed, merged, and .. accidentally unmerged. This PR is to get it back in to delivery.

File: groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java (hunk 1 above)
Comment (neilcsmith-net) [reply groovy/groovy.support/src/org/netbeans/modules/groovy/support/debug/GroovySmartStepping.java:97]: @lbownik this code was already reviewed, merged, and .. accidentally unmerged. This PR is to get it back in to delivery.

**Comment by ebarboni** [comment]
merging as previously reviewed.

**Pull request merged**